lib/libkvm/kvm_amd64.c

```diff
@@ -123,16 +124,43 @@ _amd64_initvtop(kvm_t *kd)
 		if (_kvm_read_core_phdrs(kd, &kd->vmst->phnum,
 		    &kd->vmst->phdr) == -1)
 			return (-1);
+
+		for (size_t i = 0; i < kd->vmst->phnum; i++) {
+			if (kd->vmst->phdr[i].p_type == PT_DUMP_DELTA) {
+				/* Account for the 2M hole at KERNBASE. */
+				kernphys = kd->vmst->phdr[i].p_paddr -
+				    kd->vmst->phdr[i].p_align;
+				kernbase = kd->vmst->phdr[i].p_vaddr;
+
+				found = 1;
+				break;
+			}
+		}
 	}
 
-	nl[0].n_name = "kernbase";
-	nl[1].n_name = 0;
+	if (found == 0) {
+		nl[0].n_name = "kernbase";
+		nl[1].n_name = 0;
 
-	if (kvm_nlist2(kd, nl) != 0) {
-		_kvm_err(kd, kd->program, "bad namelist - no kernbase");
-		return (-1);
+		if (kvm_nlist2(kd, nl) != 0) {
+			_kvm_err(kd, kd->program, "bad namelist - no kernbase");
+			return (-1);
+		}
+
+		nl[0].n_name = "kernphys";
+		nl[1].n_name = 0;
+
+		/* XXX
+		 * Relocatable kernels can still be loaded at 2M.
```

Comment:
Kernels loaded with copy_staging enabled in the loader (which isn't the default) will still be at 2M. There doesn't seem to be a way to detect if they were actually loaded there though.

The only way to read such dumps seems to be to read the dump anyways, at the risk of reading garbage.